docs/rewriter/design.md

```diff
@@ -0,0 +1,130 @@
+the replacement function into one function. (Currently they are different.)
+
+We can then define a rewrite rule as below:
+```py
+gemm_fusion_rule = RewriteRule (gemm_patttern, gemm_replacement, gemm_condition)
+```
+
+We have a couple of design choices with regards to some of the details of
+the above API.
+
+1. Using a _FunctionProto_ or _ONNX Script_ : We can represent a pattern
+such as the above one using a _FunctionProto_, authored using _ONNX Script_.
+
+2. We can use a custom pattern authoring language, similar to ONNX Script.
+The current implementation supports one version of such a language.
+
+Option 1 has the advantage that we reuse the existing implementation,
+and users already familiar with ONNX Script can directly use it.
+However, this approach has some limitations. For example, in the
+pattern we may not care about a required-attribute's value. If
+ONNX Script complains about a missing required-attribute, that would
+be a problem when specifying a pattern.
+
+Option 2 allows us to have a more extensible mechanism, customized to
+the case of patterns. For example, it is easy to allow users to specify
+that we are looking for attribute `alpha` to have a value of `0.16325`
+within some error tolerance limits. (Note that we can push such conditions
+into the guard/condition function too. So, this specific example is more
+of a syntactic-sugar convenience.) Another interesting use-case, which
+comes up in practice, is when we want to do some wild-card matching against
+a op/function name ... eg., if we want to match against a call to a
+function (produced by the pytorch exporter) with a name that has `Silu` as
+a prefix.
+
+## IR and Builder for Patterns and Rewrite-Rules
+
+We already have a couple of approaches to pattern-matching based rewrite-rules:
+the original one (restricted to single-output patterns, but supporting other
+features such as a mechanism to bind variables to specific nodes of the
+pattern, and using them to check for extra-conditions before applying a
+rewrite-rule, and allowing for more efficient match-and-replace), and a
+newer extension to support multiple-output patterns, but not integrated
+with the other mechanisms.
+
+We can integrate and simplify these approaches, including the different options
+for describing patterns mentioned above, by unifying the classes defining
+the patterns and rewrite-rules (into a pattern IR) and by providing different
+factory (constructor) methods for patterns and rewrite-rules. Specifically,
+these methods will be the developer-facing APIs for expressing rewrite-rules.
+Hence, it is important for us to consolidate these APIs to avoid potential
+confusion among users writing rewrite-rules. (The implementations of the
+pattern-matching can vary, depending on the complexity of the pattern,
+such as whether it is single-output or multiple-output, etc.)
+
+Proposal:
+
+* The rewrite-rule constructor takes two parameters, one to specify the
```

Comment:
I suggest keeping this extra condition parameter even if it may return True by default. Some verifications are about shapes or types. For example, it is easy to look for nodes Expand and then check that they can be removed if they expand a tensor to the same size.